### PR TITLE
EVM Error format 

### PR DESCRIPTION
**Pull request by thomas-nguy**
Closes: #247

## Description

Return Geth error format for EVM failure

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/types/errors.go [hunk 1]
```diff
@@ -75,12 +78,38 @@
...
-	reason, errUnpack := abi.UnpackRevert(revertReason)
+func NewExecErrorWithReason(revertReason []byte, vmError string) *revertError {
+	var result []byte
+	result = nil
+	if vmError == vm.ErrExecutionReverted.Error() {
+		result = common.CopyBytes(revertReason)
+	}
+
+	reason, errUnpack := abi.UnpackRevert(result)
+	err := errors.New("execution reverted")
```
Comment (yihuang) [inline x/evm/types/errors.go:89]: There are other types of errors, e.g. out of gas?

Comment (thomas-nguy) [reply x/evm/types/errors.go:89]: hmm I could not find anything special in geth. Is out of gas including in exection reverted?

Comment (yihuang) [reply x/evm/types/errors.go:89]: https://github.com/ethereum/go-ethereum/blob/master/core/vm/errors.go#L31
there are several possible cases for the `VmError`.

Comment (thomas-nguy) [reply x/evm/types/errors.go:84]: I see, you are right, I have missread the if condition

**Comment by thomas-nguy** [comment]
Currently, on transaction revert

Ethermint
```
{
 <     "jsonrpc": "2.0",
 <     "id": 1627276957344,
 <     "error": {
 <       "code": 3,
 <       "message": "execution reverted: COUNTER_TOO_LOW",
 <       "data": "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000f434f554e5445525f544f4f5f4c4f570000000000000000000000000000000000"
 <     }
 <   }
 
 ```
 
 
 Ganache
 
 ```
  <   {
 <     "id": 1627277502538,
 <     "jsonrpc": "2.0",
 <     "error": {
 <       "message": "VM Exception while processing transaction: revert COUNTER_TOO_LOW",
 <       "code": -32000,
 <       "data": {
 <         "0x90264de254689f1d4e7f8670cd97f60d9bc803874fdecb34d249bd1cc3ca823a": {
 <           "error": "revert",
 <           "program_counter": 346,
 <           "return": "0x08c379a00000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000f434f554e5445525f544f4f5f4c4f570000000000000000000000000000000000",
 <           "reason": "COUNTER_TOO_LOW"
 <         },
 <         "stack": "c: VM Exception while processing transaction: revert COUNTER_TOO_LOW\n    at Function.c.fromResults (/Users/thomasnguy/.config/yarn/global/node_modules/ganache-cli/build/ganache-core.node.cli.js:4:192416)\n    at /Users/thomasnguy/.config/yarn/global/node_modules/ganache-cli/build/ganache-core.node.cli.js:42:50402",
 <         "name": "c"
 <       }
 <     }
 ```
 
 I am not sure how compliant we needs to be. It seems that there is also differences between geth and nethermid
 https://github.com/ethereum/go-ethereum/issues/21886

**Comment by leejw51crypto** [comment]
some lint issues ^^;

File: x/evm/types/errors.go [hunk 2]
```diff
@@ -75,12 +79,33 @@ var (
 
 // NewExecErrorWithReason unpacks the revert return bytes and returns a wrapped error
 // with the return reason.
-func NewExecErrorWithReason(revertReason []byte) error {
-	hexValue := hexutil.Encode(revertReason)
-	reason, errUnpack := abi.UnpackRevert(revertReason)
+func NewExecErrorWithReason(revertReason []byte) *RevertError {
```
Comment (fedekunze) [inline x/evm/types/errors.go:82]: we should create a few unit tests for this with some of the example inputs

Comment (thomas-nguy) [reply x/evm/types/errors.go:82]: test added

**Review by fedekunze — APPROVED** [review]
LGTM, utACK

**Pull request merged**